File: 2015/day_01/src/main.rs
use std::fs::read_to_string;

const INPUT_FILE: &str = "input.txt";

const OPEN: char = '(';
const CLOSE: char = ')';

fn part_one(input: &str) {
    let mut floor = 0;
    for char in input.trim().chars() {
        match char {
            OPEN => floor += 1,
            CLOSE => floor -= 1,
            _ => {}
        }
    }
    println!("Resulting floor is {}", floor);
}

fn part_two(input: &str) {
    let mut floor = 0;
    let mut basement = 0;
    for (i, char) in input.trim().chars().enumerate() {
        match char {
            OPEN => floor += 1,
            CLOSE => floor -= 1,
            _ => {}
        }
        if floor == -1 {
            basement = i + 1;
            break;
        }
    }
    println!("Santa first enters the basement at position {}", basement);
}

fn main() {
    let input = read_to_string(INPUT_FILE).unwrap();
    part_one(&input);
    part_two(&input);
}

File: 2015/day_01/main.c
#include <stdio.h>
#include <stdbool.h>

const char INPUT[]  = "input.txt";

int main() {
	FILE *fptr = fopen(INPUT, "r");

	char ch;

	int floor = 0;
	int counter = 0;
	bool reached_basement = false;

	while ((ch = fgetc(fptr)) != EOF){
		counter += 1;
		if (ch == '(') {
			floor += 1;
		} else if (ch == ')') {
			floor -= 1;
			if (reached_basement == false && floor == -1) {
				printf("First reached basement at position: %d\n", counter); 
				reached_basement = true;
			}
			
		}
	}
	printf("Final floor: %d\n", floor); 
	fclose(fptr);
}
